pywr/notebook/save_graph.js

```
@@ -4,54 +4,59 @@ require.config({paths: {d3: 'https://d3js.org/d3.v5.min'}});
                    .range([100, -100])
                    .domain([0, height]);
-    
-    const filename = {{ filename }};
 
-    if ({{filetype}} == "model_json"){
-        for (let i = 0; i < model_data.nodes.length; i++){
-            let node_name = model_data.nodes[i].name;
-            let node_data = node_data_obj[node_name]
-            if (node_data.fixed){
+    const output_data = ["Node name,Fixed,x,y"];
+    // loop through model nodes and get postions
+    for (let i = 0; i < nodes.length; i++){
+        
+        let node_data = nodes[i];
+
+        if (!(node_data.fixed) && !(save_unfixed)) {
+            // If node is unfixed and unfixed node positions are not being saved move to next node
+            continue
+        }
+
+        if (filetype == "json"){
+            if ("position" in model_data["nodes"][i]){
+                // ensure that any geographic position are not overwritten
+                model_data["nodes"][i]["position"]["schematic"] = [posX(node_data.x), posY(node_data.y)]
+            } else {
                 model_data["nodes"][i]["position"] = {"schematic": [posX(node_data.x), posY(node_data.y)]}
             }
-        }
-        download(filename, JSON.stringify(model_data));
-    } else {
-        let output_data = ["Node name,Fixed,x,y"];
-        for (let i = 0; i < model_data.nodes.length; i++){
-            let node_name = model_data.nodes[i].name;
-            let node_data = node_data_obj[node_name]
+        } else if (filetype == "csv") {
             let position_data = [
-                node_name,
+                node_data.name,
                 node_data.fixed,
                 posX(node_data.x),
                 posY(node_data.y),
-            ]
-            output_data.push(position_data.join(","))
+            ];
+            output_data.push(position_data.join(","));
         }
-        download(filename, output_data.join("\n"));
+        // TODO append error message if filetype not recognised
```

Yes, I guess this should raise an error, but what is the JS equivalent of that?

---

I added a check for this in python instead